src/main/java/de/tum/cit/aet/artemis/programming/web/ExerciseSharingResource.java

```diff
@@ -0,0 +1,212 @@
+package de.tum.cit.aet.artemis.programming.web;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.util.Map;
+import java.util.Optional;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.ws.rs.core.UriBuilder;
+
+import org.codeability.sharing.plugins.api.ShoppingBasket;
+import org.codeability.sharing.plugins.api.util.SecretChecksumCalculator;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Profile;
+import org.springframework.core.io.InputStreamResource;
+import org.springframework.core.io.Resource;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+import de.tum.cit.aet.artemis.core.dto.SharingInfoDTO;
+import de.tum.cit.aet.artemis.programming.domain.ProgrammingExercise;
+import de.tum.cit.aet.artemis.programming.service.sharing.ExerciseSharingService;
+import de.tum.cit.aet.artemis.programming.service.sharing.ProgrammingExerciseImportFromSharingService;
+import de.tum.cit.aet.artemis.programming.service.sharing.SharingConnectorService;
+import de.tum.cit.aet.artemis.programming.service.sharing.SharingException;
+import de.tum.cit.aet.artemis.programming.service.sharing.SharingSetupInfo;
+import tech.jhipster.web.util.ResponseUtil;
+
+/**
+ * REST controller for managing the sharing of programming exercises.
+ */
+@RestController
+@RequestMapping("api/programming/sharing/")
+@Profile("sharing")
+public class ExerciseSharingResource {
+
+    /*
+     * Customized FileInputStream to delete and therefore clean up the returned files
+     */
+    static class NewFileInputStream extends FileInputStream {
```

Comment:
```suggestion
    private static class AutoDeletingFileInputStream extends FileInputStream {
```
Can be `private`?

nitpick (naming): optional, since only used within this class